ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1421 @@
+- [ ] Does the ZEP require coordination with the Operate team?
+
+-->
+
+This chapter discusses how the workflow engine will work given the proposed
+stream processing/replay.
+
+The state changes are written when processing a command or replaying an event.
+In order to align the state changes that are written on processing a command
+with the changes of replaying the resulting events, the implementation of these
+parts should be coupled or reused. For example, implementing command
+processing's logic for state changes by reusing the logic of applying state
+changes for replay of events, instead of writing the changes while processing
+the command on-the-fly.
+
+![command processor](ZEP-wf-stream-processing/event-sourcing-processor.png)
+
+Depending on the mode (processing or replay), the stream processor can skip a type of records. On processing, it reads only commands and skips events and rejections. On replay, it reads only events and skips commands and rejections. 
+
+## BPMN Examples
+
+The following examples show the sequence of records for a certain situation and explain the differences to the current behavior. 
+
+Each example is presented as a table in which:
+ - each row represents the atomic processing of a single command;
+ - rows are ordered (from top to bottom) corresponding to the order of
+   processing the commands;
+ - events and follow-up commands wrapped in square brackets (e.g.
+   `[Timer:Created]`) are optionally written (e.g. a `Deployment:Create` without
+   timer start events will not result in `Timer:Created` events).
+
+In addition to the table, notes are provided to clarify: the example; the state
+changes on processing as well as on replay; the differences to the current
+behavior; and considered design alternatives.
+
+### Deploy workflow
+
+The deployment processing has 2 interesting topics: distribution and start event
+subscriptions.
+
+#### Deployment distribution
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+Deployment:Create | Deployment:Created, [Deployment:Distribute]
+Deployment:Distribute | Deployment:Distributed
+
+Notes:
+* workflows need to be available on all partitions, Therefore, the deployment
+  needs to be distributed over all partitions. Only when processing the
+  `Deployment:Create` command on the deployment partition, the
+  `Deployment:Distribute` follow-up command is written.
+* the processing of the `Deployment:Distribute` command sends a message to the
+  other partition leaders. Each of these leaders then writes a
+  `Deployment:Create` follow-up command. Since this is written on a different
+  partition it is left out of the above table.
+* the processing of the `Deployment:Distribute` command does not require any
+  state changes on its own partition
+
+On processing:
+* the `Deployment:Create` command deploys all workflows in the deployment
+* the `Deployment:Distribute` command does not change state
```

Comment:
I agree that the description could be more detailed, but I don't think there are any major open issues here. In all of these cases the current behaviour is not any different from the proposed behaviour. It is also mentioned as future work in the future possibilities chapter. Let me answer each part individually.

>what happens if the distribution failed

it is retried by the distributor code, this is current behaviour 

>How does the receiving partition know that it doesn't have to distribute the deployment further 

The partition's stream processor knows whether it's partition is the distributing partition, this is current behaviour

>While it is distributing there is in flight state, which is not persisted

This is also currently the case, pending distributions are stored in state, but no additional measures are taken to rebuild that state after restart. Currently, if this state somehow ends up in a snapshot the pending distributions will be redistributed, and I think nothing is stopping us from keeping that.

Reply:
Are you sure that the current behavior won't change?

Today I think we reprocess the command. Since the command is doing the distribution, when we reprocess the command, we will try redistributing again.

In the new version we would only replay the events. And from what I understood the events will not redistribute the deployment. 

So replaying the new event will do less then reprocessing the current command.

If I misunderstood this, please tell me.

Reply:
> Today I think we reprocess the command. Since the command is doing the distribution, when we reprocess the command, we will try redistributing again.

Not exactly. Currently, on reprocessing, the processor adds the pending deployment to the state but it doesn't distribute it immediately (anymore). There is a scheduled task that checks for pending deployments and distributes them. This task starts after the reprocessing.

We can keep this logic for now. On replay, the processor adds the pending deployment to the state and the scheduled task distributes it.